### PR TITLE
S3 and Azure Blob Storage: Update File CDK to support document file types

### PR DESCRIPTION
**Pull request by flash1293**
## What

This PR updates the CDK for S3 and Azure Blob Storage sources.

For S3, this is moving the document file type parsing logic into the CDK.

For Azure Blob Storage, it's now supporting document file types similar to S3.

### CODE REVIEW

**Comment by flash1293** [comment]
@alafanechere All the file-based sources that support document file types will require the same build customization. For now it's not a huge lift as there aren't tons of them, but it might be nice to put this into a special spot so we won't have the same file in x places. Maybe the CDK exports the build customization (or you have another better idea)?

File: airbyte-integrations/connectors/source-azure-blob-storage/setup.py
Comment (flash1293) [inline airbyte-integrations/connectors/source-azure-blob-storage/setup.py:14]: This stuff is also included in `airbyte-cdk[file-based]`, but then there would be a version mismatch with fastavro 1.4.11 which is listed explicitly here (it's `fastavro~=1.8.0` in the file-based extra). Is this a leftover or is there a strong reason to not rely on the CDK "standard" dependencies for file based sources?

**Review by alafanechere — CHANGES_REQUESTED** [review]
Requesting changes until we find the best option to avoid `build_customization.py` code duplication

File: airbyte-integrations/connectors/source-azure-blob-storage/setup.py
Comment (alafanechere) [reply airbyte-integrations/connectors/source-azure-blob-storage/setup.py:14]: I read this comment after I submitted my first review.

Comment (alafanechere) [reply airbyte-integrations/connectors/source-azure-blob-storage/setup.py:14]: According to your PR title I would assume that these two connectors would depend on the file based CDK and the unstructured lib would be bundled in it. It's be great if connectors and CDK could agree on the `fastavro` version indeed.

File: airbyte-integrations/connectors/source-azure-blob-storage/build_customization.py
Comment (flash1293) [reply airbyte-integrations/connectors/source-azure-blob-storage/build_customization.py:1]: I agree with all of this (B for now, C in a follow-up). We could re-build the base image as part of the CDK publish action (still would require manually bumping the affected connectors but I think this would still be helpful)

Comment (alafanechere) [reply airbyte-integrations/connectors/source-azure-blob-storage/build_customization.py:1]: After thinking a bit more about it I'm not sure that C is our best long-term bet. I believe it adds up complexity and we can end up in a similar situation as we are today with `strict-encrypt` connectors. Managing variants is feasible but likely to be cumbersome. The base image will become a dependency of the file base image and compatibility, version pinning problems etc. come with it. 
As of today, I think option A - installing your new system dependencies in the base image - is the one with the lighter long-term maintenance burden.

File: airbyte-cdk/python/file_based_build_customization.py
Comment (alafanechere) [inline airbyte-cdk/python/file_based_build_customization.py:43]: Can we install specific versions of these utils to maximize build reproducibility?

File: airbyte-cdk/python/file_based_build_customization.py
Comment (alafanechere) [inline airbyte-cdk/python/file_based_build_customization.py:31]: Could you comment on which system path the nltk data will be written?

File: airbyte-cdk/python/file_based_build_customization.py
Comment (alafanechere) [inline airbyte-cdk/python/file_based_build_customization.py:58]: Something interesting to note here, which is playing against the bundling of the nltk data in the base image, is that:
- this execution would depend of the installation of `nltk` in the  base image
- the result of this execution (the file download) is dynamic. We don't control what's being downloaded exactly

This is why it makes sense for it to be a `post_connector_install` thing.
If we downloaded this data in the base image we'd have to make sure the nltk version we download this data with is the same as the one declared in the CDK... 

I'm a bit concerned by the fact that this `post_connector_install` can't guarantee reproducible results as we rely on `nltk` to decide which data is downloaded.
@flash1293 could we host snapshots of this data on one of our public bucket and make this hook download this data and mount it to the container?

File: airbyte-cdk/python/file_based_build_customization.py
Comment (alafanechere) [inline airbyte-cdk/python/file_based_build_customization.py:61]: Can this be called in a `pre_install_hook`? 
These are system dependency installation that do not depend on the python package installation if I'm not mistaken. This will speed up the build as the `apt-get install` layers will be cached. In your current implementation we'll redownload and install these tools on any code change.

File: airbyte-cdk/python/file_based_build_customization.py
Comment (alafanechere) [inline airbyte-cdk/python/file_based_build_customization.py:1]: Question about symlink and git:
Do you see the connector `build_customization.py`  (symlink) in the git diff when you modify the root `file_based_build_customization.py`.

As of today we detect connector changes based on changes in their folder with a `git diff` command. It conditionally triggers our CI and connector tests. I think we would like to keep this behavior.

In other words:
When you change `file_base_build_customization.py` do you want to release a new connector version for all the connectors using it or do you want to decouple the hook changes from the connector release?

Comment (flash1293) [reply airbyte-cdk/python/file_based_build_customization.py:1]: @alafanechere CDK changes are currently decoupled from connector changes - if a connector wants to use changed CDK stuff, the CDK needs to be published and a separate PR bumping the version needs to be opened (like what I'm doing here). It seems correct to me to extend this to the build customization script as well.

So no, this shouldn't automatically trigger a connector release. Seems like this makes the symlink approach less attractive.

File: airbyte-integrations/connectors/source-azure-blob-storage/setup.py
Comment (clnoll) [reply airbyte-integrations/connectors/source-azure-blob-storage/setup.py:14]: I'm not aware of any reason not to use the file-based CDKs dependencies.

File: airbyte-integrations/connectors/source-s3/integration_tests/spec.json
Comment (clnoll) [inline airbyte-integrations/connectors/source-s3/integration_tests/spec.json:24]:
```suggestion
          "title": "FileBasedStreamConfig",
```

File: airbyte-cdk/python/file_based_build_customization.py
Comment (flash1293) [reply airbyte-cdk/python/file_based_build_customization.py]: Moved the install to `pre_connector_install` and pinned the versions

File: airbyte-cdk/python/file_based_build_customization.py
Comment (flash1293) [reply airbyte-cdk/python/file_based_build_customization.py:74]: Inlined the nltk data index in the customization script so it points to a raw github url of that specific commit - I have a pretty high confidence this is always reproduceable (short of the maintainers deleting the repo and then we have problems anyway)

File: airbyte-cdk/python/file_based_build_customization.py
Comment (flash1293) [reply airbyte-cdk/python/file_based_build_customization.py]: Done

File: airbyte-integrations/connectors/source-azure-blob-storage/setup.py
Comment (flash1293) [reply airbyte-integrations/connectors/source-azure-blob-storage/setup.py]: Explicitly updated

File: airbyte-integrations/connectors/source-azure-blob-storage/build_customization.py
Comment (flash1293) [reply airbyte-integrations/connectors/source-azure-blob-storage/build_customization.py:1]: Not sure about the best solution, but the symlink thing seems wrong as well. For now I duplicated the file, but happy to go for a better solution

File: airbyte-integrations/connectors/source-azure-blob-storage/setup.py
Comment (flash1293) [reply airbyte-integrations/connectors/source-azure-blob-storage/setup.py]: Switched it to use the extra

File: airbyte-integrations/connectors/source-azure-blob-storage/build_customization.py
Comment (erohmensing) [reply airbyte-integrations/connectors/source-azure-blob-storage/build_customization.py:1]: Since the `build_customization.py` is just a py file that can import things, would it be possible for all of the helper methods to live in the file based CDK, imported from build_customization.py? 

Then the `pre_connector_install` and `post_connector_install` might be duplicated, but small and declarative: "before installing, install_tesseract_and_poppler`

Comment (alafanechere) [reply airbyte-integrations/connectors/source-azure-blob-storage/build_customization.py:1]: So here's what we realized could be the best approach to take with @flash1293 : 
* @flash1293 's changes mean the CDK now has system dependencies. Which is definitely a new thing.
* These system dependencies can change along the CDK version
* We should bundle the CDK system dependencies in the python connector base image
* As system dependencies can change on CDK version change we now have a coupling between CDK version used by the connector and the base image they can use.
* We should hardcode somewhere a mapping between CDK version and Base image version compatibility
* On connector build we should compare their CDK version to the base image version and fail the build if they're not compatible according to our mapping
* Connector developer will have to change the `baseImage` metadata according to this failure.

Comment (flash1293) [reply airbyte-integrations/connectors/source-azure-blob-storage/build_customization.py:1]: I think @erohmensing suggestion is interesting - this means we would need to install the CDK in the version specified in the setup.py of the connector in the python environment that building the dagger pipeline - if that's possible it might work, although it feels a little crazy.

Comment (alafanechere) [reply airbyte-integrations/connectors/source-azure-blob-storage/build_customization.py:1]: > Since the build_customization.py is just a py file that can import things, would it be possible for all of the helper methods to live in the file based CDK, imported from build_customization.py?

Then the pre_connector_install and post_connector_install might be duplicated, but small and declarative: "before installing, install_tesseract_and_poppler`

The `build_customization.py` is imported at runtime in the build process [here](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/build_image/steps/build_customization.py#L18). @erohmensing it would mean that `airbyte-ci` would depend on the CDK . As these helpers can change according to the CDK version it means we can be in a situation where we build a connector image depending on an old version of the CDK with helpers from the latest version.

**Comment by flash1293** [comment]
Thanks to @alafanechere, native dependencies are now provided by the base image. I added an acceptance test for azure to make sure the new parser actually works (this already exists for S3)

**Comment by alafanechere** [comment]
Neat, all 🟢 . Let's wait for a 👍 from my team on https://github.com/airbytehq/airbyte/pull/31929 before merging.

**Review by aaronsteers — APPROVED** [review]
Looks good! 🚀

**Review by alafanechere — APPROVED** [review]
https://github.com/airbytehq/airbyte/pull/31929 was approved
So our base image version 1.2.0 is now official

**Pull request merged**
